### DISCUSSION TITLE
Daydreaming - ideas for the future

### DISCUSSION THREAD

**Discussion by SamWindell** [floe-audio/Floe#193] [Ideas]
This is a place to put ideas. The ideas don't have to be fully thought-out at this stage. These ideas can be discarded if they don't turn out helpful.

## Right-click menu for picker menus
We currently have 'unload' option. What about 'recents' or 'favourites'.

## Pie menu
Could [pie menus](https://en.wikipedia.org/wiki/Pie_menu ) be useful on Floe's GUI instead of linear menus? [The precision radial](https://luxeengine.com/tech-precise-number-input/) is an interesting idea too.
- setting parameters to default values
- picking instruments by tag
- randomly loading instruments
![image](https://upload.wikimedia.org/wikipedia/commons/f/f6/EasyGestures_PieMenu.png)

## New minor additions
- Attack for the LFO
- Phase option for LFO
- Bypass button for effects - distinct from on/off
- Clear-all button for effects
- Tempo sync phaser frequency
- Add a sample-end parameter
- Add an alternative monophonic option that doesn't re-trigger the note if there is one playing
- Add the option to change the curves of Attack, Decay and Release
- Add a check for updates button

## Add support for other audio file types
We already support FLAC and WAV. Could Wavpack be added too? What about lossy formats: MP3, OGG?

## GUI ideas
- Add icons for filter types instead of just text
- Add ability to drag menus like a slider
- Add sustain pedal (CC64) rocker to represent it's current state

## Presets for effects rack and layers
Add the ability to save and load presets for the entire effects rack and the entire set of layers. This would allow mixing and matching different sets of sound-generators (the layers) with sound-shapers (the effects rack)

## Algorithmic legato simulation
Could we take inspiration from FrozenPlain's Arctic Strings Kontakt instrument and algorithmically simulate legato? Legato time effected by play rate and things, and a slider for portamento time - maybe even velocity controlled too

## Undo & Redo
- Undo/redo for randomly selecting presets/instruments
- Undo/redo for parameters
- Support clap's undo extension

## Detect type of sound for better tagging
Can we analyse the audio files to determine their character: pluck, sustain, bright, dark, etc. And use this information alongside the manually added tags of an instrument.

## Add the ability to parse SFZ/SoundFont/Decent Sampler files as well as our Lua format
Alternatively, could the solution for converting existing library be an external tool like this? https://github.com/git-moss/ConvertWithMoss

## Modulation system
Replace our simplistic LFOs with a full modulation system

## MIDI workflow enhancements
We currently have 'MIDI learn' but it's not very fleshed out. Could we have an interface for viewing all MIDI mappings and switch between/save profiles. 

## Screenshot generation util
It would be great to be able to script the taking of screenshots so that we can more easily keep product images up-to-date as Floe develops. Perhaps the standalone program could receive CLI arguments to load particular presets and open particular pages on the GUI and then export it as a PNG. This could then be scripted using bash

## Add benchmarking system
- Add a CLI interface for running benchmarks of Floe's audio processing
- Use hyprefine and [Bencher](https://bencher.dev/hyperfine/) to continuously benchmark

## Support QOI image format
Likely faster decoding than PNG which would be useful for backgrounds/icons. https://github.com/phoboslab/qoi/tree/master

## CI report improvements
With the ntfy notifications, it would be good to get a list of failed checks rather than just overall success or failure.

Should all of our CI just save results to json (upload-artifact) and then the final Github Actions workflow job collates all these json files (download-artifact) and uses jq to present the information however we like?

Additionally:

A dev dashboard like TigerBeetle does would be useful for Floe.

https://tigerbeetle.github.io/tigerbeetle/

Currently seeing the results of various checks and tests isn't great on the Actions tag of Github. Additionally, it would be good to graph things like benchmarks, executable size, lines of code, etc.

We could follow Tigerbeetle's template where the data is pushed to a separate github repo and then to github pages.

**Comment by pawel-ulita** [comment]
Recently, I've been evaluating Floe by attempting to convert some Kontakt libraries, mostly because of Linux compatibility.  I've generally enjoyed the process - as a programmer, working with code comes naturally to me. Equally importantly, Floe looks good, is open source and runs pretty well on Linux. Thanks for making it available!

I'd like to share my wish list for Floe. The features below would either simplify the process of creating Floe libraries or enable the creation of more expressive libraries. 

1. Convenient support for multiple microphone positions
I imagine it working by assigning multiple samples to each region, with each sample belonging to a separate microphone position. The UI would then have controls for each microphone position. Currently, it is possible to simulate this by defining multiple instruments. This method works, but has obvious drawbacks.

2. Support for multiple dimensions of timbre
A clear example of this would be controlling the dynamics and vibrato of a violin section separately.

3. Legato
This is particularly important for acoustic instruments. While a basic 'scripted' legato can be enabled through the monophonic mode, this is obviously not enough (as you mentioned above). At the same time, the ability to implement true legato would be crucial for more complex instruments. [Here](https://decentsampler-developers-guide.readthedocs.io/en/latest/topic-how-to-implement-true-legato.html) is an example of how true legato can be implemented in the DecentSampler libraries.

4. Engine inspectability
Ideally, I'd like to be able to see a dump of what the engine understood based on the Lua script. This would help me ensure that the samples are mapped to the correct keys, determine what happened to the properties that I didn't specify explicitly, and so on.

5. The ability to specify more things as part of the instrument definition that are only accessible through the UI or a preset
A major example would be the velocity-to-volume curve. For certain instruments, such as a violin section playing a long note, I would like to specify that the velocity does not affect the volume. Currently, this can only be modified as part of a preset.

I'd like to hear your thoughts on these ideas. Do you think these features are in line with your vision for the future of Floe?

**Reply by SamWindell** [reply]
It's wonderful to hear you've been trying the sample library development features. I like your suggestions a lot. I'll just share my quick thoughts on each:

1. This is a good idea. For [Dulcitone](https://www.frozenplain.com/product/dulcitone) I had separate instruments for mic positions which works pretty well, but certainly an integrated feature would be a smoother experience.
2. Interesting, not thought of this. I think the timbre options in the Lua need a bit more thinking through in general. For this use-case with violin dynamics and vibrato, is this just there would be 3 sets of samples to crossfade through? Soft dynamics, loud dynamics, and vibrato?
3. Yes this should be added for sure.
4. Great idea. I want to complete Floe's 'standalone' target. So that you can just run it as a executable in the terminal and get the output of the program (including any print() statements in Lua). Your 'engine dump' is a good idea on top of this as a quick and effective way to get information without having to develop a whole 'inspector' UI.
5. This makes sense too. Like with the loop settings you can specify what level of control you want to be exclusive to the Lua, or let the user override on the GUI.

Thanks for trying Floe.

**Reply by pawel-ulita** [reply]
Thanks for the response! Good to hear things are more or less aligned with what I had in mind.

To clarify point 2: In a simple case involving two dimensions (vibrato and dynamics), each of which has two levels, there would be a total of four sets of samples.

For example:
- no vibrato, low dynamics;
- no vibrato, high dynamics;
- moderate vibrato, low dynamics;
- moderate vibrato, high dynamics.

Then, if you want to go from no vibrato to moderate vibrato, you crossfade between 2 groups of layers, with the dynamics set to a specific level as well.

**Reply by SamWindell** [reply]
Ok understood, I'll consider for that when I enhance the timbre crossfade features.
